src/cp/slipharden.cxx

```diff
@@ -937,6 +937,302 @@ void SimpleLinearHardening::consistency(Lattice & L) const
+    X_s_(params.get_parameter<double>("X_s")),
+    varprefix_(params.get_parameter<std::string>("varprefix")), 
+    twinprefix_(params.get_parameter<std::string>("twinprefix"))
+{ 
+  // Shouldn't tau_0 be 24 and C_st be 12?
+  // Long term you should fix the /2
+  if (C_st_->n() != tau_0_.size()/2) {
+    throw std::invalid_argument("Twinning interaction matrix and initial strength sizes do not agree!");
+  }
+  
+  varnames_.resize(size());
+  for (size_t i = 0; i < size(); i++) {
+	if (i < size()/2){ // You'll need to fix this at some point
+      varnames_[i] = varprefix_+std::to_string(i);
+	} else{
+	  varnames_[i] = twinprefix_+std::to_string(i);	
+	}
+    
+  }
+  init_cache_();
+}
+
+std::string LANLTiModel::type()
+{
+  return "LANLTiModel";
+}
+
+std::unique_ptr<NEMLObject> LANLTiModel::initialize(ParameterSet & params)
+{
+  return neml::make_unique<LANLTiModel>(params);
+}
+
+ParameterSet LANLTiModel::parameters()
+{
+  ParameterSet pset(LANLTiModel::type());
+
+  pset.add_parameter<std::vector<NEMLObject>>("tau_0");
+  pset.add_parameter<NEMLObject>("C_st");
+  pset.add_parameter<std::vector<NEMLObject>>("mu");
+  pset.add_parameter<std::vector<NEMLObject>>("k1");
+  pset.add_parameter<std::vector<NEMLObject>>("k2");
+  pset.add_optional_parameter<double>("X_s", 0.9);
+  pset.add_optional_parameter<std::string>("varprefix", 
+                                           std::string("rho"));
+  pset.add_optional_parameter<std::string>("twinprefix", 
+                                           std::string("slip"));
+
+  return pset;
+}
+
+std::vector<std::string> LANLTiModel::varnames() const
+{
+  return varnames_;
+}
+
+void LANLTiModel::set_varnames(std::vector<std::string> vars)
+{
+  varnames_ = vars;
+  init_cache_();
+}
+
+void LANLTiModel::populate_history(History & history) const
+{
+  for (auto vn : varnames_) {
+    history.add<double>(vn);
+  }
+}
+
+void LANLTiModel::init_history(History & history) const
+{
+  size_t i = 0;
+  for (auto vn : varnames_) {
+    history.get<double>(vn) = 1.0e-6;
+    i++;
+  }
+}
+
+double LANLTiModel::hist_to_tau(size_t g, size_t i, 
+                                           const History & history,
+                                           Lattice & L,
+                                           double T, const History & fixed) const
+{
+  consistency(L);
+  
+  Lattice::SlipType stype = L.slip_type(g,i);	
+  if (stype == Lattice::SlipType::Slip) {
+    return X_s_ * L.burgers(g,i) * mu_[L.flat(g,i)]->value(T) 
+			    * std::sqrt(history.get<double>(varnames_[L.flat(g,i)])) 
+			    + tau_0_[L.flat(g,i)]->value(T);
+	} 
+	else {
+	double v = 0;
+	for (size_t g2 = 0; g2 < L.ngroup(); g2++) {
+	  for (size_t i2 = 0; i2 < L.nslip(g2); i2++) {
+	    size_t k2 = L.flat(g2,i2);
+	    Lattice::SlipType otype = L.slip_type(g2,i2);
+	    if (otype == Lattice::SlipType::Slip){
+		  v += (*C_st_)(L.flat(g,i)-size()/2,k2) * L.burgers(g2,i2)
```

Comment:
Don't worry about this now, but this is probably one area where we could improve the efficiency.